uefi/src/main.rs

```diff
@@ -130,173 +182,203 @@ fn main_inner(image: Handle, mut st: SystemTable<Boot>) -> Status {
+        return None;
+    }
+    let loaded_image = loaded_image.unwrap();
+    let loaded_image = loaded_image.deref();
+
+    let device_handle = loaded_image.device();
+
+    let device_path = unsafe {
+        this.open_protocol::<DevicePath>(
+            OpenProtocolParams {
+                handle: device_handle,
+                agent: image,
+                controller: None,
+            },
+            OpenProtocolAttributes::Exclusive,
+        )
+    };
+    if device_path.is_err() {
+        log::error!("Failed to open protocol DevicePath");
+        return None;
+    }
+    Some(device_path.unwrap())
+}
 
-        let fs_handle = this
-            .locate_device_path::<SimpleFileSystem>(&mut device_path)
-            .ok()?;
+fn locate_and_open_protocol<P: ProtocolPointer>(
+    image: Handle,
+    st: &SystemTable<Boot>,
+) -> Option<ScopedProtocol<P>> {
+    let this = st.boot_services();
+    let mut device_path = open_device_path_protocol(image, st)?;
+    let mut device_path = device_path.deref();
+
+    let fs_handle = this.locate_device_path::<P>(&mut device_path);
+    if fs_handle.is_err() {
+        log::error!("Failed to open device path");
+        return None;
+    }
+
+    let fs_handle = fs_handle.unwrap();
 
-        this.open_protocol::<SimpleFileSystem>(
+    let opened_handle = unsafe {
+        this.open_protocol::<P>(
             OpenProtocolParams {
                 handle: fs_handle,
                 agent: image,
                 controller: None,
             },
             OpenProtocolAttributes::Exclusive,
         )
+    };
+
+    if opened_handle.is_err() {
+        log::error!("Failed to open protocol {}", core::any::type_name::<P>());
+        return None;
     }
-    .unwrap();
-    let file_system = unsafe { &mut *file_system_raw.interface.get() };
+    Some(opened_handle.unwrap())
+}
+
+fn load_file_from_disk(
+    name: &str,
+    image: Handle,
+    st: &SystemTable<Boot>,
+) -> Option<&'static mut [u8]> {
+    let mut file_system_raw = locate_and_open_protocol::<SimpleFileSystem>(image, st)?;
+    let file_system = file_system_raw.deref_mut();
 
     let mut root = file_system.open_volume().unwrap();
-    let mut buf = [0; 14 * 2];
-    let filename = CStr16::from_str_with_buf("kernel-x86_64", &mut buf).unwrap();
-    let kernel_file_handle = root
-        .open(filename, FileMode::Read, FileAttribute::empty())
-        .expect("Failed to load kernel (expected file named `kernel-x86_64`)");
-    let mut kernel_file = match kernel_file_handle.into_type().unwrap() {
+    let mut buf = [0u16; 256];
+    assert!(name.len() < 256);
+    let filename = CStr16::from_str_with_buf(name.trim_end_matches('\0'), &mut buf)
+        .expect("Failed to convert string to utf16");
+
+    let file_handle_result = root.open(filename, FileMode::Read, FileAttribute::empty());
+
+    if file_handle_result.is_err() {
+        return None;
+    }
+
+    let file_handle = file_handle_result.unwrap();
```

Comment:
We can use a match here to avoid the `unwrap()`:

```suggestion
    let file_handle = match file_handle_result {
        Err(_) => return None,
        Some(handle) = handle,
    };
```